streams/src/main/java/org/apache/kafka/streams/state/internals/InMemoryTimeOrderedKeyValueBuffer.java

```diff
@@ -222,60 +230,109 @@ public void flush() {
+                                                         .putLong(bufferKey.time)
+                                                         .put(serializedContextualRecord)
+                                                         .array();
+
+        collector.send(
+            changelogTopic,
+            key,
+            timeAndContextualRecord,
+            V_1_CHANGELOG_HEADERS,
+            partition,
+            null,
+            KEY_SERIALIZER,
+            VALUE_SERIALIZER
+        );
+    }
+
+    private void logTombstone(final Bytes key) {
+        collector.send(changelogTopic,
+                       key,
+                       null,
+                       null,
+                       partition,
+                       null,
+                       KEY_SERIALIZER,
+                       VALUE_SERIALIZER
+        );
+    }
+
     private void restoreBatch(final Collection<ConsumerRecord<byte[], byte[]>> batch) {
         for (final ConsumerRecord<byte[], byte[]> record : batch) {
             final Bytes key = Bytes.wrap(record.key());
             if (record.value() == null) {
                 // This was a tombstone. Delete the record.
                 final BufferKey bufferKey = index.remove(key);
                 if (bufferKey != null) {
-                    sortedMap.remove(bufferKey);
+                    final ContextualRecord removed = sortedMap.remove(bufferKey);
+                    if (removed != null) {
+                        memBufferSize -= computeRecordSize(bufferKey.key, removed);
+                    }
+                    if (bufferKey.time == minTimestamp) {
+                        final BufferKey leastKey = sortedMap.firstKey();
+                        minTimestamp = leastKey != null ? leastKey.time : Long.MAX_VALUE;
+                    }
+                }
+
+                if (record.partition() != partition) {
+                    throw new IllegalStateException(
+                        String.format(
+                            "record partition [%d] is being restored by the wrong suppress partition [%d]",
+                            record.partition(),
+                            partition
+                        )
+                    );
                 }
             } else {
                 final ByteBuffer timeAndValue = ByteBuffer.wrap(record.value());
                 final long time = timeAndValue.getLong();
                 final byte[] value = new byte[record.value().length - 8];
                 timeAndValue.get(value);
+                if (record.headers() == null || record.headers().lastHeader("v") == null) {
+                    cleanPut(time, key,
+                             new ContextualRecord(
+                                 value,
+                                 new ProcessorRecordContext(
+                                     record.timestamp(),
```

Comment:
Unless I'm missing something aren't we restoring the record with the changelog `offset`, `topic` etc?

Reply:
Yeah, this is the (incorrect) behavior of the current code. Luckily, this will only happen for the very small window of emitting records that are currently buffered when you upgrade Streams to "this" version. Once those are flushed out, the buffer will always be sending/receiving "v1" changelog messages that correctly encode the upstream topic/offset.

I figured the best way to handle this transition is to just preserve the existing behavior for "old format" messages, rather than try to do something clever.

Reply:
We're setting the record metadata incorrectly here, which is actually the bug I'm fixing with the new format. It's just that there's apparently nothing better to do that continue handling the old changelog format the same way, since the format itself is missing the information we need.